From c172aab871ee89e472c4852f5684f50ca19a670d Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Mon, 20 Oct 2014 14:32:06 -0400
Subject: [fs] nfs: Fix a bogus warning in nfs_generic_pgio

Message-id: <1413815526-10854-3-git-send-email-steved@redhat.com>
Patchwork-id: 97679
O-Subject: [RHEL7.1 PATCH 2/2] NFS: Fix a bogus warning in nfs_generic_pgio
Bugzilla: 1150227
RH-Acked-by: Ian Kent <ikent@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

It is OK for pageused == pagecount in the loop, as long as we don't add
another entry to the *pages array. Move the test so that it only triggers
in that case.

Reported-by: Steve Dickson <SteveD@redhat.com>
Fixes: bba5c1887a92 (nfs: disallow duplicate pages in pgio page vectors)
Cc: Weston Andros Adamson <dros@primarydata.com>
Cc: stable@vger.kernel.org # 3.16.x
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: b8fb9c30f25e45dab5d2cd310ab6913b6861d00f
Bugzilla: 1150227
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/pagelist.c b/fs/nfs/pagelist.c
index 8aafc3d..52be1e8 100644
--- a/fs/nfs/pagelist.c
+++ b/fs/nfs/pagelist.c
@@ -753,12 +753,11 @@ int nfs_generic_pgio(struct nfs_pageio_descriptor *desc,
   nfs_list_remove_request(req);
   nfs_list_add_request(req, &hdr->pages);
 
-  if (WARN_ON_ONCE(pageused >= pagecount))
-   return nfs_pgio_error(desc, hdr);
-
   if (!last_page || last_page != req->wb_page) {
-   *pages++ = last_page = req->wb_page;
    pageused++;
+   if (pageused > pagecount)
+    break;
+   *pages++ = last_page = req->wb_page;
   }
  }
  if (WARN_ON_ONCE(pageused != pagecount))
-- 
1.7.1